File: kore/src/main/scala/org/kframework/POSet.scala
// Copyright (c) Runtime Verification, Inc. All Rights Reserved.
package org.kframework

import collection._
import java.util
import org.kframework.utils.errorsystem.KEMException
import scala.annotation.tailrec

/**
 * A partially ordered set based on an initial set of direct relations.
 */
class POSet[T](val directRelations: Set[(T, T)]) extends Serializable {
  // convert the input set of relations to Map form for performance
  private val directRelationsMap: Map[T, Set[T]] =
    directRelations.groupBy(_._1).mapValues(_.map(_._2).toSet).map(identity)

  lazy val elements: Set[T] = directRelations.flatMap(a => Set(a._1, a._2))

  lazy val sortedElements: scala.collection.immutable.List[T] =
    TopologicalSort.tsort(directRelations).toList

  /**
   * Internal private method. Computes the transitive closer of the initial relations. It also
   * checks for cycles during construction and throws an exception if it finds any.
   *
   * The implementation is simple. It links each element to the successors of its successors.
   *
   * TODO: there may be a more efficient algorithm (low priority)
   */
  @tailrec
  private def transitiveClosure(relations: Map[T, Set[T]]): Map[T, Set[T]] = {
    val newRelations = relations.map { case (start, succ) =>
      val newSucc = succ.flatMap(relations.getOrElse(_, Set()))
      if (newSucc.contains(start))
        constructAndThrowCycleException(start, start, Seq())
      (start, succ | newSucc)
    }
    if (relations != newRelations) transitiveClosure(newRelations) else relations
  }

  /**
   * Recursive method constructing and throwing and the cycle exception.
   *
   * @param start
   *   (or tail) element to look for when constructing the cycle
   * @param current
   *   element
   * @param path
   *   so far
   */
  private def constructAndThrowCycleException(start: T, current: T, path: Seq[T]): Unit = {
    val currentPath = path :+ current
    val succs       = directRelationsMap.getOrElse(current, Set())
    if (succs.contains(start)) {
      throw KEMException.compilerError(
        "Illegal circular relation: " + (currentPath :+ start).mkString(" < ")
      )
    }
    succs.foreach(constructAndThrowCycleException(start, _, currentPath))
  }

  /**
   * All the relations of the POSet, including the transitive ones.
   *
   * Concretely, a map from each element of the poset to the set of elements greater than it.
   */
  val relations: Map[T, Set[T]] = transitiveClosure(directRelationsMap)

  /**
   * A map from each element of the poset to the set of elements less than it.
   */
  lazy val relationsOp: Map[T, Set[T]] =
    relations
      .toSet[(T, Set[T])]
      .flatMap { case (x, ys) => ys.map(_ -> x) }
      .groupBy(_._1)
      .mapValues(_.map(_._2))

  def <(x: T, y: T): Boolean = relations.get(x).exists(_.contains(y))
  def >(x: T, y: T): Boolean = relations.get(y).exists(_.contains(x))
  def ~(x: T, y: T): Boolean = <(x, y) || <(y, x)

  /**
   * Returns true if x < y
   */
  def lessThan(x: T, y: T): Boolean         = <(x, y)
  def lessThanEq(x: T, y: T): Boolean       = x == y || <(x, y)
  def directlyLessThan(x: T, y: T): Boolean = directRelationsMap.get(x).exists(_.contains(y))

  /**
   * Returns true if y < x
   */
  def greaterThan(x: T, y: T): Boolean         = >(x, y)
  def greaterThanEq(x: T, y: T): Boolean       = x == y || >(x, y)
  def directlyGreaterThan(x: T, y: T): Boolean = directRelationsMap.get(y).exists(_.contains(x))

  /**
   * Returns true if y < x or y < x
   */
  def inSomeRelation(x: T, y: T): Boolean   = this.~(x, y)
  def inSomeRelationEq(x: T, y: T): Boolean = x == y || this.~(x, y)

  /**
   * Return the set of all upper bounds of the input.
   */
  def upperBounds(sorts: Iterable[T]): Set[T] =
    if (sorts.isEmpty) elements else POSet.upperBounds(sorts, relations)

  def upperBounds(sorts: util.Collection[T]): util.Set[T] =
    Collections.mutable(upperBounds(Collections.immutable(sorts)))

  /**
   * Return the set of all lower bounds of the input.
   */
  def lowerBounds(sorts: Iterable[T]): Set[T] =
    if (sorts.isEmpty) elements else POSet.upperBounds(sorts, relationsOp)

  def lowerBounds(sorts: util.Collection[T]): util.Set[T] =
    Collections.mutable(lowerBounds(Collections.immutable(sorts)))

  lazy val minimalElements: Set[T] = minimal(elements)

  lazy val maximalElements: Set[T] = maximal(elements)

  lazy val maximum: Option[T] =
    if (maximalElements.size == 1) Some(maximalElements.head) else None

  lazy val minimum: Option[T] =
    if (minimalElements.size == 1) Some(minimalElements.head) else None

  lazy val asOrdering: Ordering[T] = (x: T, y: T) =>
    if (lessThanEq(x, y)) -1 else if (lessThanEq(y, x)) 1 else 0

  /**
   * Return the subset of items from the argument which are not less than any other item.
   */
  def maximal(sorts: Iterable[T]): Set[T] =
    sorts.filter(s1 => !sorts.exists(s2 => lessThan(s1, s2))).toSet

  def maximal(sorts: util.Collection[T]): util.Set[T] =
    Collections.mutable(maximal(Collections.immutable(sorts)))

  /**
   * Return the subset of items from the argument which are not greater than any other item.
   */
  def minimal(sorts: Iterable[T]): Set[T] =
    sorts.filter(s1 => !sorts.exists(s2 => >(s1, s2))).toSet

  def minimal(sorts: util.Collection[T]): util.Set[T] =
    Collections.mutable(minimal(Collections.immutable(sorts)))

  override def toString: String =
    "POSet(" + relations
      .flatMap { case (from, tos) => tos.map(to => from + "<" + to) }
      .mkString(",") + ")"

  override def hashCode: Int = relations.hashCode()

  override def equals(that: Any): Boolean = that match {
    case that: POSet[_] => relations == that.relations
    case _              => false
  }
}

object POSet {
  def apply[T](relations: (T, T)*) = new POSet(relations.toSet)
  def apply[T](s: Set[(T, T)])     = new POSet(s)

  /**
   * Import this for Scala syntactic sugar.
   */
  implicit class PO[T](x: T)(implicit val po: POSet[T]) {
    def <(y: T): Boolean = po.<(x, y)
    def >(y: T): Boolean = po.>(x, y)
  }

  /**
   * Return the set of all elements which are greater than or equal to each element of the input,
   * using the provided relations map. Input must be non-empty.
   */
  private def upperBounds[T](sorts: Iterable[T], relations: Map[T, Set[T]]): Set[T] =
    sorts.map(s => relations.getOrElse(s, Set.empty) + s).reduce((a, b) => a & b)
}

File: k-frontend/src/main/java/org/kframework/POSet.java
// Copyright (c) Runtime Verification, Inc. All Rights Reserved.
package org.kframework;

import java.io.Serializable;
import java.util.ArrayDeque;
import java.util.Arrays;
import java.util.Comparator;
import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;
import java.util.Optional;
import java.util.Set;
import java.util.stream.Collectors;
import java.util.stream.Stream;
import org.apache.commons.lang3.tuple.Pair;
import org.kframework.utils.Lazy;
import org.kframework.utils.errorsystem.KEMException;
import org.pcollections.Empty;
import org.pcollections.PStack;

public class POSet<T> implements Serializable {
  private final Set<Pair<T, T>> directRelations;
  private final Map<T, Set<T>> directRelationsMap;
  private final Map<T, Set<T>> relations;

  @SafeVarargs
  public POSet(Pair<T, T>... directRelations) {
    this(Arrays.stream(directRelations).collect(Collectors.toSet()));
  }

  public POSet(scala.collection.Set<scala.Tuple2<T, T>> directRelations) {
    this(
        Collections.stream(directRelations)
            .map(t -> Pair.of(t._1, t._2))
            .collect(Collectors.toSet()));
  }

  public POSet(Set<Pair<T, T>> directRelations) {
    this.directRelations = directRelations;
    this.directRelationsMap = directRelationsMap(directRelations);
    this.relations = transitiveClosure(directRelationsMap);
  }

  private static <T> Map<T, Set<T>> directRelationsMap(Set<Pair<T, T>> directRelations) {
    return directRelations.stream()
        .collect(Collectors.groupingBy(Pair::getLeft))
        .entrySet()
        .stream()
        .collect(
            Collectors.toMap(
                Map.Entry::getKey,
                e -> e.getValue().stream().map(Pair::getRight).collect(Collectors.toSet())));
  }

  private Map<T, Set<T>> transitiveClosure(Map<T, Set<T>> relations) {
    boolean changed = false;
    Map<T, Set<T>> newRelations = new HashMap<>();
    for (Map.Entry<T, Set<T>> rel : relations.entrySet()) {
      T start = rel.getKey();
      Set<T> succ = rel.getValue();
      Set<T> newSucc =
          succ.stream()
              .flatMap(s -> relations.getOrDefault(s, new HashSet<>()).stream())
              .collect(Collectors.toSet());
      if (newSucc.contains(start)) {
        constructAndThrowCycleException(start, start, Empty.stack());
      }

      newSucc.addAll(succ);
      if (newSucc.size() > succ.size()) {
        changed = true;
      }
      newRelations.put(start, newSucc);
    }
    if (changed) {
      return transitiveClosure(newRelations);
    }
    return relations;
  }

  private void constructAndThrowCycleException(T start, T current, PStack<T> path) {
    PStack<T> currentPath = path.plus(current);
    Set<T> succs = directRelationsMap.getOrDefault(current, new HashSet<>());
    if (succs.contains(start)) {
      // we built the path in reverse, so use plusAll to reverse it back
      PStack<T> finalPath = Empty.<T>stack().plusAll(currentPath.plus(start));
      throw KEMException.compilerError(
          "Illegal circular relation: "
              + finalPath.stream().map(Object::toString).collect(Collectors.joining(" < ")));
    }
    succs.forEach(s -> constructAndThrowCycleException(start, s, currentPath));
  }

  public Map<T, Set<T>> relations() {
    return relations;
  }

  private Set<T> computeElements() {
    return directRelations.stream()
        .flatMap(p -> Stream.of(p.getLeft(), p.getRight()))
        .collect(Collectors.toSet());
  }

  private final Lazy<Set<T>> elementsLazy = new Lazy<>(this::computeElements);

  public Set<T> elements() {
    return elementsLazy.get();
  }

  private java.util.List<T> computeSortedElements() {
    Optional<Stream<T>> topological = DirectedGraph.topologicalSort(directRelations);
    // We already checked for cycles during construction, so the sort should succeed
    assert topological.isPresent();
    return topological.get().toList();
  }

  private final Lazy<java.util.List<T>> sortedElementsLazy =
      new Lazy<>(this::computeSortedElements);

  public java.util.List<T> sortedElements() {
    return sortedElementsLazy.get();
  }

  private Map<T, Set<T>> computeRelationsOp() {
    return relations.entrySet().stream()
        .flatMap(e -> e.getValue().stream().map(val -> Pair.of(val, e.getKey())))
        .collect(Collectors.groupingBy(Pair::getLeft))
        .entrySet()
        .stream()
        .collect(
            Collectors.toMap(
                Map.Entry::getKey,
                e -> e.getValue().stream().map(Pair::getRight).collect(Collectors.toSet())));
  }

  private final Lazy<Map<T, Set<T>>> relationsOpLazy = new Lazy<>(this::computeRelationsOp);

  public Map<T, Set<T>> relationsOp() {
    return relationsOpLazy.get();
  }

  private Set<T> computeMinimalElements() {
    return minimal(elements());
  }

  private final Lazy<Set<T>> minimalElementsLazy = new Lazy<>(this::computeMinimalElements);

  public Set<T> minimalElements() {
    return minimalElementsLazy.get();
  }

  private Set<T> computeMaximalElements() {
    return maximal(elements());
  }

  private final Lazy<Set<T>> maximalElementsLazy = new Lazy<>(this::computeMaximalElements);

  public Set<T> maximalElements() {
    return maximalElementsLazy.get();
  }

  private Optional<T> computeMinimum() {
    Set<T> minimalElems = minimalElements();
    if (minimalElems.size() == 1) {
      return Optional.of(minimalElems.iterator().next());
    }
    return Optional.empty();
  }

  private final Lazy<Optional<T>> minimumLazy = new Lazy<>(this::computeMinimum);

  public Optional<T> minimum() {
    return minimumLazy.get();
  }

  private Optional<T> computeMaximum() {
    Set<T> maximalElems = maximalElements();
    if (maximalElems.size() == 1) {
      return Optional.of(maximalElems.iterator().next());
    }
    return Optional.empty();
  }

  private final Lazy<Optional<T>> maximumLazy = new Lazy<>(this::computeMaximum);

  public Optional<T> maximum() {
    return maximumLazy.get();
  }

  private Map<T, Integer> computeConnectedComponents() {
    var data = new HashMap<T, Integer>();
    int nextComponent = 0;

    for (var node : elements()) {
      var queue = new ArrayDeque<T>();
      queue.push(node);

      var anyChange = false;
      while (!queue.isEmpty()) {
        var cur = queue.pop();
        if (data.containsKey(cur)) {
          continue;
        }

        data.put(cur, nextComponent);
        anyChange = true;

        for (var next : relations().getOrDefault(cur, Set.of())) {
          queue.addFirst(next);
        }

        for (var next : relationsOp().getOrDefault(cur, Set.of())) {
          queue.addFirst(next);
        }
      }

      if (anyChange) {
        nextComponent++;
      }
    }

    return data;
  }

  private final Lazy<Map<T, Integer>> connectedComponentsLazy =
      new Lazy<>(this::computeConnectedComponents);

  /**
   * Compute a connected component identifier for each element in this POSet.
   *
   * <p>Two elements share a connected component identifier if there is some sequence of edges (in
   * any direction) that connect them. Note that being in the same connected component does not
   * imply that two elements are related to one another in their POSet; if `A <: B` and `A <: C`
   * then `B` and `C` are in the same connected component but are not in any ordering relation with
   * each other.
   *
   * <p>Note that this mapping is computed lazily when this method is first called; it uses a DFS
   * over the POSet graph internally and has time complexity of `O(V + E)`. Subsequent calls will
   * not recompute the mapping and therefore take constant time.
   *
   * @return A lazily initialized map from elements to an integer in the range [0, N) indicating
   *     which connected component that element belongs to, where N is the number of connected
   *     components.
   */
  public Map<T, Integer> connectedComponents() {
    return connectedComponentsLazy.get();
  }

  public boolean lessThan(T x, T y) {
    return relations.getOrDefault(x, Set.of()).contains(y);
  }

  public boolean lessThanEq(T x, T y) {
    return x.equals(y) || lessThan(x, y);
  }

  public boolean directlyLessThan(T x, T y) {
    return directRelationsMap.getOrDefault(x, Set.of()).contains(y);
  }

  public boolean greaterThan(T x, T y) {
    return relations.getOrDefault(y, Set.of()).contains(x);
  }

  public boolean greaterThanEq(T x, T y) {
    return x.equals(y) || greaterThan(x, y);
  }

  public boolean directlyGreaterThan(T x, T y) {
    return directRelationsMap.getOrDefault(y, Set.of()).contains(x);
  }

  public boolean inSomeRelation(T x, T y) {
    return lessThan(x, y) || greaterThan(x, y);
  }

  public boolean inSomeRelationEq(T x, T y) {
    return x.equals(y) || inSomeRelation(x, y);
  }

  private enum BoundType {
    UPPER,
    LOWER
  }

  private Set<T> bounds(BoundType boundType, Iterable<T> elems) {
    Map<T, Set<T>> relationsOrRelationsOp =
        switch (boundType) {
          case UPPER -> relations;
          case LOWER -> relationsOp();
        };
    return Collections.streamIter(elems)
        .map(
            e -> {
              Set<T> bounds = new HashSet<>(relationsOrRelationsOp.getOrDefault(e, Set.of()));
              bounds.add(e);
              return bounds;
            })
        .reduce(
            (a, b) -> {
              a.retainAll(b);
              return a;
            })
        .orElse(elements());
  }

  public Set<T> upperBounds(Iterable<T> elems) {
    return bounds(BoundType.UPPER, elems);
  }

  public Set<T> lowerBounds(Iterable<T> elems) {
    return bounds(BoundType.LOWER, elems);
  }

  public Set<T> minimal(Iterable<T> elems) {
    return Collections.streamIter(elems)
        .filter(s1 -> Collections.streamIter(elems).noneMatch(s2 -> greaterThan(s1, s2)))
        .collect(Collectors.toSet());
  }

  public Set<T> maximal(Iterable<T> elems) {
    return Collections.streamIter(elems)
        .filter(s1 -> Collections.streamIter(elems).noneMatch(s2 -> lessThan(s1, s2)))
        .collect(Collectors.toSet());
  }

  public Comparator<T> asComparator() {
    return (T x, T y) -> lessThan(x, y) ? -1 : (greaterThan(x, y) ? 1 : 0);
  }

  @Override
  public String toString() {
    return "POSet("
        + relations.entrySet().stream()
            .flatMap(e -> e.getValue().stream().map(val -> e.getKey() + "<" + val))
            .collect(Collectors.joining(","))
        + ")";
  }

  @Override
  public int hashCode() {
    return relations.hashCode();
  }

  @Override
  public boolean equals(Object obj) {
    if (obj instanceof POSet<?> po) {
      return relations.equals(po.relations);
    }
    return false;
  }
}
